vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c

```diff
@@ -1083,11 +752,294 @@ BTStatus_t prvBTSendResponse( uint16_t usConnId,
+    struct ble_gatt_svc_def * pSvc = NULL;
+
+    if( serviceCnt < MAX_SERVICES )
+    {
+        pSvc = &espServices[ serviceCnt ];
+        afrServices[ serviceCnt ] = pxService;
+    }
+    else
+    {
+        xStatus = eBTStatusNoMem;
+    }
+
+    /* Fill in service field. After that start creating characteristics and descriptors individually. */
+    if( xStatus == eBTStatusSuccess )
+    {
+        espServices[ serviceCnt + 1 ].type = BLE_GATT_SVC_TYPE_END;
+
+        if( pxService->pxBLEAttributes[ 0 ].xAttributeType == eBTDbPrimaryService )
+        {
+            pSvc->type = BLE_GATT_SVC_TYPE_PRIMARY;
+        }
+        else if( pxService->pxBLEAttributes[ 0 ].xAttributeType == eBTDbPrimaryService )
+        {
+            pSvc->type = BLE_GATT_SVC_TYPE_SECONDARY;
+        }
+        else
+        {
+            configPRINTF( ( "Invalid service type \n" ) );
+            xStatus = eBTStatusFail;
+        }
+    }
+
+    if( xStatus == eBTStatusSuccess )
+    {
+        pSvc->uuid = prvCopytoESPUUID( &pxService->pxBLEAttributes[ 0 ].xServiceUUID, NULL );
+
+        if( pSvc->uuid == NULL )
+        {
+            configPRINTF( ( "Could not allocate memory for service UUID \n" ) );
+            xStatus = eBTStatusNoMem;
+        }
+        else
+        {
+            pCharacteristics = pvPortCalloc( prvCountCharacteristics( pxService ) + 1, sizeof( struct ble_gatt_chr_def ) );
+
+            if( pCharacteristics == NULL )
+            {
+                configPRINTF( ( "Could not allocate memory for  characteristic array \n" ) );
+                prvCleanupService( pxService, pSvc );
+                xStatus = eBTStatusNoMem;
+            }
+
+            pSvc->characteristics = pCharacteristics;
+            pSvc->includes = NULL;
+            charCount = 0;
+            dscrCount = 0;
+        }
+    }
+
+    /* Create characteristics and descriptor individually .*/
+    if( xStatus == eBTStatusSuccess )
+    {
+        for( index = 0; index < pxService->xNumberOfAttributes; index++ )
+        {
+            switch( pxService->pxBLEAttributes[ index ].xAttributeType )
+            {
+                case eBTDbCharacteristic:
+                    /* Allocate memory for UUID and copies it. */
+                    pCharacteristics[ charCount ].uuid = prvCopytoESPUUID( &pxService->pxBLEAttributes[ index ].xCharacteristic.xUuid, NULL );
+                    pCharacteristics[ charCount ].arg = ( void * ) pxService;
+                    pCharacteristics[ charCount ].access_cb = prvGATTCharAccessCb;
+                    pCharacteristics[ charCount ].flags = prvAFRToESPCharPerm( pxService->pxBLEAttributes[ index ].xCharacteristic.xProperties, pxService->pxBLEAttributes[ index ].xCharacteristic.xPermissions );
+
+                    /* Allocate memory for descriptors. */
+                    pCharacteristics[ charCount ].descriptors = pvPortCalloc( prvCountDescriptor( pxService, index + 1 ) + 1, sizeof( struct ble_gatt_dsc_def ) );
```

Comment:
Also this pvPortCalloc(). If allocation fails would previous allocations need to be cleaned up?

Reply:
Yes, but this one seems ok. It gets cleaned here: 
            if( xStatus != eBTStatusSuccess )
            {
                configPRINTF( ( "Failed to allocate memory during Attribute creation\n" ) );
                prvCleanupService( pxService, pSvc );
                break;
            }